Allow to add extra accelerators to wxMenuItem

kosh543 opened the pull request:
These accelerators will not be shown in wxMenuItem label, but still will work.

kosh543 committed:
Allow to add extra accelerators to wxMenuItem

These accelerators will not be shown in wxMenuItem label, but still will work.

Co-authored-by: VZ <vz-github@zeitlins.org>

vadz commented on src/xrc/xh_menu.cpp:
```diff
@@ -101,6 +114,17 @@ wxObject *wxMenuXmlHandler::DoCreateResource()
+            if (!extra_accels.empty())
+            {
+                const int entries_size = extra_accels.size();
+                for (wxString extra_accel : extra_accels)
```
Unfortunately we still can't use C++11 in wx 3.1, please rewrite the loop using indices (and then we'll rewrite it back like this after 3.2.0...).

vadz commented on src/xrc/xh_menu.cpp:
```diff
@@ -79,6 +80,18 @@ wxObject *wxMenuXmlHandler::DoCreateResource()
+            wxVector<wxString> extra_accels;
+            if (HasParam(wxT("extra-accels")))
+            {
+                wxXmlNode* extra_accels_node = GetParamNode(wxT("extra-accels"));
+                wxXmlNode* extra_accel_node = extra_accels_node->GetChildren();
```
A few things here:

- We use `camelCase` in wx.
- Local variables should be `const` when they don't change.
- Having two variables differing by a single letter should be avoided if possible.

So I'd like to change this to something like this instead:
```suggestion
                wxXmlNode* const extraAccelsNode = GetParamNode(wxT("extra-accels"));
                wxXmlNode* node = extraAccelsNode->GetChildren();
```

vadz commented on src/xrc/xh_menu.cpp:
```diff
@@ -101,6 +114,17 @@ wxObject *wxMenuXmlHandler::DoCreateResource()
+                for (wxString extra_accel : extra_accels)
+                {
+                    wxAcceleratorEntry entry;
+                    if (entry.FromString(extra_accel))
```
Should we give an error if this fails? Or is this never supposed to fail here? If it isn't, maybe we should store `wxAcceleratorEntries` directly, i.e. `wxVector<wxAcceleratorEntry>`?

vadz commented on docs/doxygen/overviews/xrc_format.h:
```diff
@@ -518,6 +518,28 @@ and can be one of the following "sub-properties":
+Defines a list of wxMenuItem's extra accelerators.
+
+The extra-accels property element is a "composite" element:
+it contains one or more <accel> "sub-properties":
```
I think angle brackets need to be escaped (please run Doxygen locally to check):
```suggestion
it contains one or more @c \<accel\> "sub-properties":
```

vadz commented on docs/doxygen/overviews/xrc_format.h:
```diff
@@ -1496,6 +1518,9 @@ wxMenuItem objects support the following properties:
      Item's accelerator (default: none).}
+@row3col{extra-accels, @ref overview_xrcformat_type_extra_accels,
+     List of item's extra accelerators. Such accelerators will not be shown
+     in item's label, but still will work. (default: none).}
```
```suggestion
     in item's label, but still will work. (default: none).
     This property is only supported since wxWidgets 3.1.6.}
```

vadz commented on include/wx/gtk/menuitem.h:
```diff
@@ -39,6 +39,10 @@ class WXDLLIMPEXP_CORE wxMenuItem : public wxMenuItemBase
     void SetGtkLabel();
 
+#if wxUSE_ACCEL
+    void SetGtkExtraAccels();
```
We use `GTK` prefix by convention for wxGTK-specific methods that are not part of the public API (also `MSW`, `OSX`, ...).
```suggestion
    void GTKSetExtraAccels();
```

vadz commented on include/wx/menuitem.h:
```diff
@@ -163,6 +170,10 @@ class WXDLLIMPEXP_CORE wxMenuItemBase : public wxObject
     bool          m_isEnabled;      // is enabled?
 
+#if wxUSE_ACCEL
+    wxVector<wxString> m_extra_accels; // extra accels will work, but won't be shown in wxMenuItem title
```
Let's use camel case for the members:
```suggestion
    wxVector<wxString> m_extraAccels; // extra accels will work, but won't be shown in wxMenuItem title
```

(also see the suggestion below for using a vector of accelerator entries directly).

vadz commented on include/wx/menuitem.h:
```diff
@@ -122,6 +122,13 @@ class WXDLLIMPEXP_CORE wxMenuItemBase : public wxObject
+    // add the accel to extra accels list
+    void AddExtraAccel(wxAcceleratorEntry *accel);
+
+    wxVector<wxString> GetExtraAccels() const { return m_extra_accels; }
```
Doesn't matter much for vectors of this size, but it should still arguably be returned by const reference rather than copied.

vadz commented on include/wx/menuitem.h:
```diff
@@ -122,6 +122,13 @@ class WXDLLIMPEXP_CORE wxMenuItemBase : public wxObject
     virtual void SetAccel(wxAcceleratorEntry *accel);
+
+    // add the accel to extra accels list
+    void AddExtraAccel(wxAcceleratorEntry *accel);
```
What's the reason for taking a pointer here? I.e. why not
```suggestion
    void AddExtraAccel(const wxAcceleratorEntry& accel);
```
?

vadz commented on interface/wx/menuitem.h:
```diff
@@ -567,6 +567,25 @@ class wxMenuItem : public wxObject
+       but still will work.
+
+       @onlyfor{wxmsw,wxgtk}
+    */
```
```suggestion

        @since 3.1.6
    */
```

vadz commented on interface/wx/menuitem.h:
```diff
@@ -567,6 +567,25 @@ class wxMenuItem : public wxObject
     virtual void SetAccel(wxAcceleratorEntry *accel);
 
+    /**
+       Add the accel to extra accels list for this item,
+       such accels will not be shown in item's label,
+       but still will work.
```
```suggestion
       Add an extra accelerator for this menu item.
       
       Additional accelerators are not shown in the item's label,
       but still will trigger the menu command when pressed.
       
       They can be useful to let multiple keys be used as accelerators
       for the same command, e.g. @c WXK_ADD and @c WXK_NUMPAD_ADD.
```

vadz commented on interface/wx/menuitem.h:
```diff
@@ -567,6 +567,25 @@ class wxMenuItem : public wxObject
+    wxVector<wxString> GetExtraAccels() const;
+
+    /**
+       Clear the accels list.
```
```suggestion
       Clear the extra accelerators list.
       
       This doesn't affect the main item accelerator (if any).
       
       @since 3.1.6
```

vadz commented on interface/wx/menuitem.h:
```diff
@@ -567,6 +567,25 @@ class wxMenuItem : public wxObject
+    */
+    void AddExtraAccel(wxAcceleratorEntry *accel);
+
+    /**
+       Return the accels list from this item.
+    */
+    wxVector<wxString> GetExtraAccels() const;
```
Not sure if we really need to make this one public, is it actually useful for the application code?

If we do, then it should have `@since 3.1.6` too.

vadz commented on samples/xrc/myframe.cpp:
```diff
@@ -285,6 +286,14 @@ void MyFrame::OnUncenteredToolOrMenuCommand(wxCommandEvent& WXUNUSED(event))
+void MyFrame::OnMultipleAccels(wxCommandEvent& WXUNUSED(event))
+{
+    wxString msg;
+    msg.Printf( "You can open this dialog also with 'Ctrl-Q' and 'Shift-Ctrl-W' accelerators.");
```
But actually it has `Ctrl-W` as main accelerator too, so I'd rather say
```suggestion
    msg.Printf( "You can open this dialog with any of `Ctrl-W` (main), 'Ctrl-Q' or 'Shift-Ctrl-W' (extra) accelerators.");
```

vadz commented on src/common/menucmn.cpp:
```diff
@@ -296,6 +296,14 @@ void wxMenuItemBase::SetAccel(wxAcceleratorEntry *accel)
+{
+    if ( accel )
+    {
+        m_extra_accels.push_back(accel->ToString());
```
I wonder if we shouldn't use `ToRawString()` here to avoid any problems due to localization.

vadz commented on src/common/menucmn.cpp:
```diff
@@ -296,6 +296,14 @@ void wxMenuItemBase::SetAccel(wxAcceleratorEntry *accel)
 
+void wxMenuItemBase::AddExtraAccel(wxAcceleratorEntry *accel)
+{
+    if ( accel )
```
If we need to pass a pointer here, this should be a `wxCHECK`, as I don't see why would you ever want to intentionally pass a null pointer here. But better would be to avoid pointers in the first place.

vadz commented on src/msw/menu.cpp:
```diff
@@ -225,29 +225,36 @@ void wxMenu::UpdateAccel(wxMenuItem *item)
-        else
+
+        // add extra accels
+        for ( wxString extra_accel_str : item->GetExtraAccels() )
```
This one also needs to be rewritten in C++98.

vadz commented on src/msw/menu.cpp:
```diff
@@ -225,29 +225,36 @@ void wxMenu::UpdateAccel(wxMenuItem *item)
-            delete m_accels[n];
-            if ( accel )
-                m_accels[n] = accel;
+            wxAcceleratorEntry *extra_accel = new wxAcceleratorEntry;
+            if ( extra_accel->FromString(extra_accel_str) )
```
```suggestion
            if ( wxAcceleratorEntry* const extraAccel = wxAcceleratorEntry::Create(extra_accel_str) )
```

vadz commented on src/msw/menu.cpp:
```diff
@@ -225,29 +225,36 @@ void wxMenu::UpdateAccel(wxMenuItem *item)
+            }
             else
-                m_accels.RemoveAt(n);
+            {
+                // failed to create accel
+                delete extra_accel;
+            }
```
```suggestion
```

vadz reviewed (dismissed):
Sorry, didn't have time to look at wxGTK code yet, but I wanted to already publish this.

Globally it looks fine, but there are a lot of minor comments to fix and I'd also like to have a non-interactive unit test for this somewhere under `tests/menu`, ideally including `wxUIActionSimulator` checking that the extra accels actually do work.

TIA!

vadz commented on samples/xrc/myframe.cpp:
```diff
@@ -285,6 +286,14 @@ void MyFrame::OnUncenteredToolOrMenuCommand(wxCommandEvent& WXUNUSED(event))
+void MyFrame::OnMultipleAccels(wxCommandEvent& WXUNUSED(event))
+{
+    wxString msg;
+    msg.Printf( "You can open this dialog with any of `Ctrl-W` (main), 'Ctrl-Q' or 'Shift-Ctrl-W' (extra) accelerators.");
```
Sorry, my fault, I've somehow mistyped the quotes, but they should be the same for all of them, of course:
```suggestion
    msg.Printf( "You can open this dialog with any of 'Ctrl-W' (main), 'Ctrl-Q' or 'Shift-Ctrl-W' (extra) accelerators.");
```

vadz commented on src/gtk/menu.cpp:
```diff
@@ -678,6 +679,35 @@ void wxMenuItem::SetGtkLabel()
+#if wxUSE_ACCEL
+void wxMenuItem::SetGtkExtraAccels()
+{
+    for (wxString extra_accel : m_extra_accels)
```
Another C++11 loop needs to be rewritten.

vadz commented on src/gtk/menu.cpp:
```diff
@@ -678,6 +679,35 @@ void wxMenuItem::SetGtkLabel()
+    for (wxString extra_accel : m_extra_accels)
+    {
+        wxAcceleratorEntry entry;
+        if ( entry.FromString(extra_accel) )
```
Same question as elsewhere about what should happen if this fails.

vadz commented on src/gtk/menu.cpp:
```diff
@@ -678,6 +679,35 @@ void wxMenuItem::SetGtkLabel()
+                    wxGTK_CONV_SYS(extra_hot_key),
+                    &extra_accel_key, &extra_accel_mods
+                );
+                if ( gtk_accelerator_valid(extra_accel_key, extra_accel_mods) )
```
We need to log a message similar to the one in `wxGetGtkAccel()` (maybe it can refactored and reused?) here to avoid the situation in which accelerators are simply silently ignored and people don't know why they don't work.

vadz commented on src/gtk/menu.cpp:
```diff
@@ -1313,13 +1346,20 @@ static wxString GetGtkHotKey( const wxMenuItem& item )
+{
+    wxAcceleratorEntry *accel = item.GetAccel();
+    wxString hotkey = GetGtkHotKeyFromAccel(accel);
+    if ( accel )
```
Test is not really needed.

But I'd actually do just
```cpp
wxScopedPtr<wxAcceleratorEntry> accel = item.GetAccel();
return GetGtkHotKeyFromAccel(accel.get());
```

vadz commented on src/gtk/menu.cpp:
```diff
@@ -1313,13 +1346,20 @@ static wxString GetGtkHotKey( const wxMenuItem& item )
     return hotkey;
 }
 
+static wxString GetGtkHotKey( const wxMenuItem& item )
```
In fact, looking at the code, I'm not even sure if we still need this function, it's used in a single place only and it looks like it might be simpler to just inline it into `wxGetGtkAccel()`

vadz reviewed (commented):
Just a small update after looking at wxGTK code too (spoiler: no important comments there neither).

vadz commented on src/msw/menu.cpp:
```diff
@@ -225,29 +225,30 @@ void wxMenu::UpdateAccel(wxMenuItem *item)
+
+        // add extra accels
+        const wxVector<wxAcceleratorEntry>& extraAccelsVector = item->GetExtraAccels();
+        const int extraAccelsSize = extraAccelsVector.size();
```
Doesn't this provoke a warning from MSVC about `size_t` to `int` conversion? We need to use `size_t` indices for the loops...

vadz reviewed (commented):
Great, thanks, looks good to me.

Let's see if you can make it work under Mac, but even if you can't, this would be already useful to have for MSW/GTK.

But the tests would be nice to have too, of course.

vadz commented on src/osx/cocoa/menuitem.mm:
```diff
@@ -270,6 +270,15 @@ void Hide( bool hide ) wxOVERRIDE
+    void SetAllowsKeyEquivalentWhenHidden( bool allow ) wxOVERRIDE
+    {
+        // setAllowsKeyEquivalentWhenHidden is available since macOS 10.13
+        if (@available(macOS 10.13, *))
```
We use `WX_IS_MACOS_AVAILABLE()` from `wx/osx/private/available.h` for such checks. I'm not sure if the rationale there is still valid, but I think it might be, and it would be better to use the macro here too if only for consistency.

vadz commented on src/osx/cocoa/menuitem.mm:
```diff
@@ -270,6 +270,15 @@ void Hide( bool hide ) wxOVERRIDE
+        if (@available(macOS 10.13, *))
+            [m_osxMenuItem setAllowsKeyEquivalentWhenHidden:allow ];
+        else
+            wxLogDebug("wxMenuItemCocoaImpl::setAllowsKeyEquivalentWhenHidden not supported under OS X < 10.13");
```
To be really useful, this message should say something about "Extra accelerators not being supported under macOS < 10.13", as the application programmer will probably have no idea why are we trying to do it in the first place.

In fact, I think we should just avoid doing anything at all under macOS < 10.13, i.e. ignore the extra accelerators completely, if we can't make them work anyhow.

vadz commented on tests/menu/menu.cpp:
```diff
@@ -172,7 +173,18 @@ void MenuTestCase::CreateFrame()
+        extraAccels->AddExtraAccel(entry);
+    if (entry.FromString("Ctrl-W"))
+        extraAccels->SetAccel(&entry);
+#if defined(__WXOSX_COCOA__)
+    m_itemCount += 2;
+#endif
```
We need to hide this difference inside wxOSX itself.

vadz committed:
fixup! Allow to add extra accelerators to wxMenuItem

vadz committed:
Make GetGtkHotKeyFromAccel() argument const

No real changes, just don't require a non-const pointer in a function
which doesn't change its argument.

vadz committed:
Add GtkAccel helper abstracting operations with GTK accelerators

This simple struct combines both the accelerator code and flags and is
simpler to work with than two different variables.

It also encapsulates the check for validity of the accelerator.

vadz committed:
Update GTK accelerators when extra accelerators change

Previously the extra accelerators were only taken into account when the
item was appended to the menu, but not if they were changed later.

Check for this in the overridden AddExtraAccel() and ClearExtraAccels()
and update the GTK accelerators as well if the item is already attached
to the menu.

vadz commented:
I'm closing this in favour of #2588, which is essentially identical (I've just added some extra tests).
